### PR TITLE
Improve spatial models and add diffuse models

### PR DESCRIPTION
**Pull request by joleroi**
* [x] Make all spatial model work with numpy arrays
* [x] Use ``gammapy.maps`` in TemplateModel
* [x] Accept path as input to ``read`` methods in ``gammapy.maps``

### CODE REVIEW

File: gammapy/image/models/tests/test_models.py [hunk 1]
```diff
@@ -74,11 +87,22 @@ def test_skyshell2D():
     desired = 0.0002976757280439522 / u.deg ** 2
```
Comment (joleroi) [inline gammapy/image/models/tests/test_models.py]: @cdeil The test value changed quite a bit here, you wrote the old ``TemplateModel`` class. Any ideas why this is?

File: gammapy/image/models/tests/test_models.py [hunk 2]
```diff
@@ -74,11 +87,22 @@ def test_skyshell2D():
...
 def test_template2d():
     filename = ('$GAMMAPY_EXTRA/datasets/catalogs/fermi/Extended_archive_v18'
                 '/Templates/HESSJ1841-055.fits')
     template = SkyTemplate2D.read(filename)
-    assert_allclose(template(26.7, 0), 1.1553735159851262)
+    lon = 26.7 * u.deg
+    lat = 0 * u.deg
+    actual = template(lon, lat)[0]
+    desired = 0.00017506744188722223 / u.deg ** 2
+    #desired = 1.1553735159851262 / u.deg ** 2
```
Comment (joleroi) [inline gammapy/image/models/tests/test_models.py]: @cdeil The test value changed quite a bit here. You wrote the previous version of the TemplateModel class. Any idea why this is?

**Comment by cdeil** [comment]
@joleroi - Thanks!
I've continued a bit here. E.g. the point source model still had an if. And the return values from evaluate were sometimes in deg-2, sometimes sr-1, sometimes without unit. I've now changed to go to rad at the start, do the model evaluation without units, and return in sr-1.  Merging now.

Several things here are preliminary and need more thought: should we take the quantity handling out of evaluate completely? How should diffuse models be normalised and should the diffuse model map unit be taken into account or not? In init we probably should take parameter objects and defer the quantity -> parameter conversion to the Parameter class, not do it in the spatial model inits.
We can sort these things out in future PRs.

File: gammapy/image/models/new.py [hunk 3]
```diff
@@ -42,12 +41,6 @@ def __str__(self):
             ss += '\n\t'.join(covar.pformat())
         return ss
 
-    @staticmethod
-    @abc.abstractmethod
```
Comment (joleroi) [inline gammapy/image/models/new.py]: Why did you remove this? Isn't the point of having an abstract base class that you define an API that the daughter classes have to follow?

File: gammapy/image/models/tests/test_models.py [hunk 4]
```diff
@@ -67,15 +68,31 @@ def test_sky_shell():
...
-def test_sky_template():
+def test_sky_diffuse_map():
     filename = ('$GAMMAPY_EXTRA/datasets/catalogs/fermi/Extended_archive_v18'
-                '/Templates/HESSJ1841-055.fits')
-    template = SkyTemplate.read(filename)
-    lon = 26.7 * u.deg
-    lat = 0 * u.deg
-    actual = template(lon, lat)[0]
-    desired = 0.00017506744188722223 / u.deg ** 2
-    # desired = 1.1553735159851262 / u.deg ** 2
```
Comment (joleroi) [inline gammapy/image/models/tests/test_models.py]: So do you know why we cannot reproduce the old test value?

Comment (cdeil) [reply gammapy/image/models/tests/test_models.py]: I didn't have a look at the old values. I don't think it's worth, becase a) there were onlly 1-2 users of this (Axel + ?) and the behaviour wasn't well specified anyways. E.g. I think the old version was applying normalisation on init, and integrating over pixels, which are two things I think this class shouldn't do. Note that I left some TODO in the class of things I wanted to discuss with you concerning diffuse models. This is just a rough first draft. Do you have time today or later this week?

**Comment by cdeil** [comment]
I forgot to update the 3D example scripts. Done in master in:
https://github.com/gammapy/gammapy/commit/a05ffd906a8943fcbde71a38a98c826bcd37ecdb
